crates/uv-pypi-types/src/conflicts.rs

```
@@ -47,6 +55,153 @@ impl Conflicts {
+                    let mut canonical_items = FxHashSet::default();
+                    canonical_items.insert(item.clone());
+                    let node_id = graph.add_node(canonical_items);
+                    group_node_idxs.insert(group.clone(), node_id);
+                    seen.insert(group);
+                    node_conflict_items.insert(node_id, item.clone());
+                }
+            }
+        }
+
+        // Create conflict items for remaining groups and add them to the graph.
+        for group in groups.keys() {
+            if !seen.contains(group) {
+                seen.insert(group);
+                let group_conflict_item = ConflictItem {
+                    package: package.clone(),
+                    conflict: ConflictPackage::Group(group.clone()),
+                };
+                let node_id = graph.add_node(FxHashSet::default());
+                group_node_idxs.insert(group.clone(), node_id);
+                node_conflict_items.insert(node_id, Rc::new(group_conflict_item));
+            }
+        }
+
+        // Create edges representing group inclusion (with edges reversed so that
+        // included groups point to including groups).
+        for (group, specifiers) in groups {
+            let includer = group_node_idxs
+                .get(group)
+                .expect("Group should have been added to graph");
+            for specifier in specifiers {
+                if let DependencyGroupSpecifier::IncludeGroup { include_group } = specifier {
+                    let included = group_node_idxs
+                        .get(include_group)
+                        .expect("Group should have been added to graph");
+                    graph.add_edge(*included, *includer, ());
+                }
+            }
+        }
+
+        // Propagate canonical items through the graph and populate substitutions.
+        // FIXME: Have we already done cycle detection before this method was
```

Can we assume cycle detection before this point? Otherwise, we'll have to return an error

---

There's a failing test that wasn't running locally for some reason that shows we can't assume it at this point. I'll update.

---

What I've done now is to just return if there's a `Cycle` error on `toposort`. I've left a `FIXME` comment indicating  that we're currently bailing and allowing more detailed include cycle detection downstream. Is this the behavior we want? If so, I'll remove the `FIXME` tag.

---

Hmmm cc @zanieb 

One thing that strikes me here is that the cycle detection error would only kick in when a relevant conflicting group is declared? What happens if there aren't any conflicts declared? Is there cycle detection happening somewhere else?

---

Cycle detection is happening somewhere, maybe in the resolver? But that happens after this code is called. With this new version where I just return (no error) when `toposort` returns an error, then only the original conflict sets are sent along. 

But it does seem weird to silently swallow the error. The reason I don't want it propagated is because better cycle detection happens elsewhere ("there was a cycle" vs. "there was a cycle: dev -> test -> dev"). So I could either warn the user in this method or match at the caller and warn there. Do either of those options make sense?

---

I think what you're doing now is probably reasonable with a comment explaining what's going on here (basically, what you just said). @zanieb might have a different opinion though!

---

I don't know what our current behavior for cycles is. Are we talking about cycles in inclusions or cycles in the conflicts?

1. If `dev` includes `test` and `test` includes `dev` don't we have a consistent set of dependencies? (just the union of `test` and `dev`?) 

2. Similarly, if `test` is marked as conflicting with some other group `foo` then `dev` conflicts with `foo` too, the cycle doesn't matter here.

3. If `test` is marked as conflicting with `dev`, now the cycle is a problem because... they include each other and can't conflict. This applies to longer chains too, e.g., if we have `test ->  foo -> dev -> test` a conflict between `test` and `dev` is "invalid".

If we don't have test coverage for each of these cases, we should add it. In reply to

> One thing that strikes me here is that the cycle detection error would only kick in when a relevant conflicting group is declared?

This seems fine, because this form of cycle (3) is the only one that creates an unresolvable situation?

---

Unless I'm misunderstanding, cycles in dependency groups are an error. Per the PEP:

> Dependency Group Includes MUST NOT include cycles, and tools SHOULD report an error if they detect a cycle.

I think we have an `DependencyGroupCycle` error.

---

Yeah, on `main` we definitely error on cycles. That's the behavior I'm relying on when just returning from the conflict expansion method when detecting a cycle during `toposort`.

---

Oh, thanks for the correction. Maybe I was just thinking about feedback I wrote on the PEP.